lib/ops/opsservice/update.go

```diff
@@ -537,24 +537,31 @@ func (s *site) checkUpdateParameters(update *pack.PackageEnvelope, provisioner s
 }
 
 func (s *site) validateDockerConfig(updateManifest schema.Manifest) error {
+	existingDocker, err := ops.GetDockerConfig(s.service, s.key)
+	if err != nil {
+		return trace.Wrap(err)
+	}
 	docker := updateManifest.SystemOptions.DockerConfig()
 	if docker == nil {
-		// No changes
-		return nil
-	}
-
-	existingDocker := s.dockerConfig()
-	if existingDocker.IsEmpty() {
-		installOperation, err := ops.GetCompletedInstallOperation(s.key, s.service)
-		if err != nil {
-			return trace.Wrap(err)
+		// If Docker driver wasn't specified explicitly in manifest and
+		// the cluster is currently using devicemapper, the upgrade
+		// will fail b/c devicemapper is no longer working and the
+		// current behavior is to preserve the driver if it's not
+		// specified in the manifest, so check here and don't let to
+		// proceed.
+		if existingDocker.StorageDriver != constants.DockerStorageDriverDevicemapper {
+			return nil
 		}
+		return trace.BadParameter(
```

Comment:
The reason I did it this way is b/c of the current behavior that the storage driver is preserved b/w upgrades unless specified explicitly. Without this check upgrades to 5.5 would just fail. I guess we could instead just force-upgrade the driver implicitly but that would be a change in behavior.

Reply:
My 2 cents: If left unspecified, I think it's safer and cleaner UX to interpret unspecified as the recommended driver from gravitational, which seems reasonable for us to switch during the upgrade as we can't support maintaining the driver.

Also, does this alert get thrown during tele build or during upgrade? Because it's in the opseervice/update code I'm assuming during upgrade, which I don't think is ideal.

Reply:
I agree it will be a better UX. I've updated the logic to use default driver if not specified in manifest.